File: frontend/public/videos/scripts/question.py
import openai
import os
from dotenv import load_dotenv, find_dotenv
import argparse
import json
from pathlib import Path

load_dotenv(find_dotenv())
client = openai.OpenAI()
openai.api_key = os.environ.get("OPENAI_API_KEY")


def is_valid_json(json_str):
    try:
        json.loads(json_str)
        return True
    except json.JSONDecodeError:
        return False


def get_translation_question(japanese_text: str) -> dict:
    prompt = f"""
        Translate the following subtitles and create three more incorrect translations for people learning Japanese. Return it in JSON format like the following:
        If subtitles are "＜謎のペンダントを持つ女の子シータを悪者達から守り→", return this
        {{
            "translation": "Protecting the girl Sheeta who possesses a mysterious pendant from the villains",
            "incorrect_translations": [ 
                "Protecting the mysterious girl's pendant from the villains",
                "Sheeta, the girl with the strange necklace, defends the villains.",
                "To save the evil people with the pendant that the girl Sheeta owns."
            ]
        }}
        """

    print("Translating with ChatGPT:", japanese_text)
    try:
        response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[
                {
                    "role": "system",
                    "content": "You are a helpful assistant that translates Japanese subtitles.",
                },
                {"role": "user", "content": prompt},
                {"role": "user", "content": japanese_text},
            ],
            temperature=0.3,
        )
    except Exception as e:
        print(f"Error with OpenAI API: {e}")
        return {}

    content = response.choices[0].message.content
    print("Received:", content)
    if not is_valid_json(content):
        print("ChatGPT returned invalid JSON. Aborting...")
        return {}
    json_content = json.loads(content)
    if (
        "translation" not in json_content
        or "incorrect_translations" not in json_content
    ):
        print("Couldn't find necessary keys in the JSON response.")
        return {}

    return json_content


def update_json_file(file_path: Path):
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            data = json.load(f)

        japanese_text = data.get("subtitles", "")
        if not japanese_text:
            print(f"No 'subtitles' field found in {file_path}. Skipping...")
            return

        translation_data = get_translation_question(japanese_text)
        if not translation_data:
            print(f"Failed to get translation for {file_path}. Skipping...")
            return

        data.update(translation_data)

        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=4)

        print(f"Updated {file_path} successfully.")

    except Exception as e:
        print(f"Error processing {file_path}: {e}")


def main():
    parser = argparse.ArgumentParser(
        description="Recursively update JSON files with translations and incorrect options."
    )
    parser.add_argument(
        "--path",
        required=True,
        help="Path to the root directory containing JSON files.",
    )
    args = parser.parse_args()

    root_path = Path(args.path)
    if not root_path.exists():
        print(f"The path {root_path} does not exist.")
        return

    json_files = list(root_path.rglob("*.json"))
    if not json_files:
        print("No JSON files found.")
        return

    print(f"Found {len(json_files)} JSON files. Starting processing...")
    for json_file in json_files:
        update_json_file(json_file)


if __name__ == "__main__":
    main()

File: frontend/public/videos/scripts/question.ts
import { Configuration, OpenAIApi } from "openai";
import { promises as fs } from "fs";
import path from "path";
import { Command } from "commander";

const configuration = new Configuration({
	apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

async function isValidJson(jsonStr: string): Promise<boolean> {
	try {
		JSON.parse(jsonStr);
		return true;
	} catch {
		return false;
	}
}

interface TranslationResponse {
	translation: string;
	incorrect_translations: string[];
}

async function getTranslationQuestion(
	japaneseText: string,
): Promise<TranslationResponse | null> {
	const prompt = `
        Translate the following subtitles and create three more incorrect translations for people learning Japanese. Return it in JSON format like the following:
        If subtitles are "＜謎のペンダントを持つ女の子シータを悪者達から守り→", return this
        {
            "translation": "Protecting the girl Sheeta who possesses a mysterious pendant from the villains",
            "incorrect_translations": [ 
                "Protecting the mysterious girl's pendant from the villains",
                "Sheeta, the girl with the strange necklace, defends the villains.",
                "To save the evil people with the pendant that the girl Sheeta owns."
            ]
        }
    `;

	console.log("Translating with OpenAI:", japaneseText);

	try {
		const response = await openai.createChatCompletion({
			model: "gpt-3.5-turbo",
			messages: [
				{
					role: "system",
					content:
						"You are a helpful assistant that translates Japanese subtitles.",
				},
				{ role: "user", content: prompt },
				{ role: "user", content: japaneseText },
			],
			temperature: 0.3,
		});

		const content = response.data.choices[0].message.content;
		if (await isValidJson(content)) {
			const jsonContent: TranslationResponse = JSON.parse(content);
			if (jsonContent.translation && jsonContent.incorrect_translations) {
				return jsonContent;
			}
		}
		console.error("Invalid response structure:", content);
	} catch (error) {
		console.error("OpenAI API error:", error);
	}

	return null;
}

async function updateJsonFile(
	filePath: string,
	callCount: number,
	maxCalls: number,
	overwrite: boolean,
): Promise<number> {
	if (callCount >= maxCalls) {
		console.log(
			`Reached the maximum API call limit of ${maxCalls}. Skipping further calls.`,
		);
		return callCount;
	}

	try {
		const data: any = JSON.parse(await fs.readFile(filePath, "utf8"));

		if (data.translation && !overwrite) {
			console.log(
				`Skipping ${filePath} (already contains 'translation').`,
			);
			return callCount;
		}

		const japaneseText = data.subtitles;
		if (!japaneseText) {
			console.log(
				`No 'subtitles' field found in ${filePath}. Skipping...`,
			);
			return callCount;
		}

		const translationData = await getTranslationQuestion(japaneseText);
		if (!translationData) {
			console.log(
				`Failed to get translation for ${filePath}. Skipping...`,
			);
			return callCount;
		}

		Object.assign(data, translationData);
		await fs.writeFile(filePath, JSON.stringify(data, null, 4), "utf8");
		console.log(`Updated ${filePath} successfully.`);
		return callCount + 1;
	} catch (error) {
		console.error(`Error processing ${filePath}:`, error);
		return callCount;
	}
}

async function main() {
	const program = new Command();
	program
		.requiredOption(
			"--path <path>",
			"Path to the root directory containing JSON files.",
		)
		.option(
			"--max-calls <number>",
			"Maximum number of API calls to make.",
			parseInt,
			10,
		)
		.option(
			"--overwrite",
			"Overwrite existing translations if present.",
			false,
		);
	program.parse(process.argv);

	const options = program.opts();
	const rootPath = path.resolve(options.path);
	const maxCalls = options.maxCalls || 10; // Default value if not specified
	const overwrite = options.overwrite;

	try {
		const files = await fs.readdir(rootPath, { withFileTypes: true });
		const jsonFiles: string[] = [];

		for (const file of files) {
			if (file.isFile() && file.name.endsWith(".json")) {
				jsonFiles.push(path.join(rootPath, file.name));
			}
		}

		console.log(
			`Found ${jsonFiles.length} JSON files. Starting processing with a limit of ${maxCalls} API calls...`,
		);
		let callCount = 0;

		for (const jsonFile of jsonFiles) {
			callCount = await updateJsonFile(
				jsonFile,
				callCount,
				maxCalls,
				overwrite,
			);
			if (callCount >= maxCalls) break;
		}
	} catch (error) {
		console.error("Error reading directory:", error);
	}
}

main().catch((error) => console.error("Script error:", error));
